### PR TITLE
Adds MNIST example Jupyter/Colab notebook.

### PR DESCRIPTION
**Pull request by andsteing**
As part of issues #479 and #231.

For reviewing the Colab, see
https://colab.sandbox.google.com/github/andsteing/flax/blob/mnist/linen_examples/mnist/mnist.ipynb

Note: you can comment the notebook directly in this mirrored PR : andsteing#3

### CODE REVIEW

File: linen_examples/mnist/mnist.ipynb
Comment (andsteing) [inline linen_examples/mnist/mnist.ipynb:190]: We should probably stress here that data is PUBLIC when uploaded to tensorboard.dev

**Comment by avital** [comment]
I like the way this PR visualizes wrong predictions but I don't really love a few things here:
1. The whole initialization with copying the MNIST example out is not something we would expect our users to do, why are we doing it here?
2. The actual model and training code which is what people most want to tinker with isn't available here. (And there's no clear way to modify the code if you want to tinker)

I would think it's more helpful to our users to instead supply a maximally simple MNIST Colab that they can directly tinker with, maybe something like the annotated MNIST version (https://flax.readthedocs.io/en/latest/annotated_mnist.html). 

WDYT?

File: linen_examples/mnist/mnist.ipynb
Comment (andsteing) [reply linen_examples/mnist/mnist.ipynb:92]: Absolutely! Thanks for highlighting this.

**Comment by andsteing** [comment]
> I like the way this PR visualizes wrong predictions but I don't really love a few things here:
> 
> 1. The whole initialization with copying the MNIST example out is not something we would expect our users to do, why are we doing it here?

When this code is run in Jupyter, then no initialization is performed.

On the other hand, when the Colab is opened directly via the
https://colab.sandbox.google.com/github/andsteing/flax/blob/mnist/linen_examples/mnist/mnist.ipynb
link then the files in the same directory are not available (`$CWD` is in a newly created VM).

So if we want to make the notebooks run on Colab as opposed to only with Jupyter, then we need to do some setup. On the positive side, this setup should be the same for every notebook and it's only 9 LOC.

> 2. The actual model and training code which is what people most want to tinker with isn't available here. (And there's no clear way to modify the code if you want to tinker)
> 
> I would think it's more helpful to our users to instead supply a maximally simple MNIST Colab that they can directly tinker with, maybe something like the annotated MNIST version (https://flax.readthedocs.io/en/latest/annotated_mnist.html).
> 
> WDYT?

I disagree : I think tinkering with the `.py` files should be done in your favourite text editor where you have syntax highlighting etc, instead of in a Colab cell. The Colab should only contain the glue code to let you interactively run different parts of the example code and inspect the inputs/outputs.

Of course nothing prevents you from copying in code from the example into a cell and execute it from there. But I would rather avoid us making this copy and then having to keep it in sync.

Having a maximally simple MNIST Colab is a different story that we can put into the docs separately. This PR is about a model Colab that should be put into every existing `example` folder.

**Comment by avital** [comment]
> I disagree : I think tinkering with the .py files should be done in your favourite text editor where you have syntax highlighting etc, instead of in a Colab cell. The Colab should only contain the glue code to let you interactively run different parts of the example code and inspect the inputs/outputs.

OK, maybe -- I see that point. But then how would one do this with this Colab? Is there a way to edit the MNIST code locally and run it in Colab? If that's the case that's great -- and should probably be documented.

If we want people to use this from Jupyter only, then that's fine and I wouldn't call these Colabs. But that seems suboptimal for people wanting to tinker with models on GPU and TPU?

I guess I'd like to better understand the expected user flow of these notebooks

**Comment by avital** [comment]
> Having a maximally simple MNIST Colab is a different story that we can put into the docs separately. This PR is about a model Colab that should be put into every existing example folder.

Sounds good.

**Comment by andsteing** [comment]
> OK, maybe -- I see that point. But then how would one do this with this Colab? Is there a way to edit the MNIST code locally and run it in Colab? If that's the case that's great -- and should probably be documented.

Yes, you can edit it directly in Colab UI:
mnist_colab

Of course you can also develop on your local machine and then push/pull via a branch in your forked Flax repo.

> If we want people to use this from Jupyter only, then that's fine and I wouldn't call these Colabs. But that seems suboptimal for people wanting to tinker with models on GPU and TPU?

I think it's great having code run both in Jupyter and Colab, especially for easy access to accelerators.

> I guess I'd like to better understand the expected user flow of these notebooks

Should I create some new documentation for this? Maybe in `examples/README.md` or should we add additional documentation under `docs/examples/colab.rst`?

Then we could link that new documentation in the beginning of every notebook.

**Comment by avital** [comment]
Oh, I didn't know about that editing ability. But then where are those files saved? Is there a risk that with Colab kernel restart you lose your changes?

**Comment by andsteing** [comment]
Yes they're gone with the VM. We could link Drive to persist files. But then I feel the setup code becomes more cumbersome.

Maybe we should not encourage people explicitly to use the Colab UI after all (I think it's a nice feature to play around with, but unexperienced users might be confused by the VM persistence semantics).

I think the workflow should be like:
1. Develop on your local machine with your favourite editor & interact in Jupyter notebook.
2. Once you're happy with your changes and want to try something with a GPU/TPU : push to your forked flax repo & load in Colab.
3. If you really want, you could change files in Colab editor & push them back. But I feel that's getting complicated.

But maybe that's more of an advanced usecase.

I still think it's nice to have Colab because it allows people to play around with the example without installing anything.

**Comment by avital** [comment]
> I still think it's nice to have Colab because it allows people to play around with the example without installing anything.

I agree with that sentiment, but I'm not convinced (yet?) that this approach will give users that, so I'd rather we discuss this more before we add Colabs like this to all example.

**Comment by andsteing** [comment]
Reminder: Colab can be reviewed at
https://colab.sandbox.google.com/github/andsteing/flax/blob/mnist/linen_examples/mnist/mnist.ipynb
(=> Suggested actions : Inspect, clear outputs, run all, ...)

Hi Avital

Thanks for the time reviewing this thoroughly and all your valuable input!

Please see the updated notebook - I think that I addressed many of your concerns:

1. There's a link to the `annotated_mnist_example._ipynb` notebook that is precisely the MNIST example with all the code inlined. I agree this is an important usecase, but I don't think it makes sense to replicate that code here as well. BTW : any idea why the file ends in `._ipynb` instead of the more standard `.ipynb`?
2. I explain the "Flax Notebook Workflow" prominently in the beginning of the notebook. (Note that everything works just fine in Jupyter[lab] as well)
3. The editor now opens `mnist_lib.py` automatically and the code can directly be modified. Very easy. I strongly believe that it is preferable having a horizontal split on all but 90-dgree-rotated screens. Also, it has the advantage that no cell needs to be re-run after updating the code.
4. Persisting all code in your personal Google Drive is as simple as selecting `yes` in the drop down.
5. Tensorboard is shown inline (additional to the previous "upload to tensorboard.dev").
6. The Colab now defaults to the GPU runtime and does a simple loop over a hyper parameter for illustration purposes.

Note that I do NOT think that we should have an identical notebook for every example, but we should rather think carefully in every case what brings most benefit to the user. I think that the notebooks will be a main figurehead for the examples - especially with the Colab links figuring prominently on the `README.md`.

Some minor open questions:

a. Do you agree `mount_gdrive` should default to `yes`? The current setup has the advantage to run end-to-end without any user interaction (authenticating for Drive requires manually copying a auth token).
b. Similar question and reasoning for `upload_data` dropdown.
c. Do you agree we should commit the notebook *with* the outputs? It makes it rather large (~10M) mainly because of TensorBoard - that won't display when viewed on Github ([link](https://github.com/google/flax/blob/d662da5948c8c8c671220d18608375b0a23eecf9/linen_examples/mnist/mnist.ipynb)), but it shows fine with on Colab ([link](https://colab.sandbox.google.com/github/andsteing/flax/blob/mnist/linen_examples/mnist/mnist.ipynb)). Also, including the outputs might discourage users from actually running it (?).

Cheers, Andreas

**Comment by avital** [comment]
Thanks @andsteing -- this really looks much better!

Honestly this flow where you set up local editing with Google Drive with one click, and then allowing editing example code (or Flax core code) without running any cells is really innovative. It's better than any other Jupyter or Colab flow that I know of (when I use Jupyter[Lab] locally I typically need to go back and forth between an IDE/Editor and the notebook tab)

I think this template really does support all of the "user journeys" we discussed -- both for someone new to Flax trying to learn and tinker, as well as advanced users wanting to iterate on accelerators either locally or in Colab. And we could use this to also make custom Colabs for fine-tuning and other interesting downstream tasks.

This PR LGTM. I suspect there are a few small improvements we could make as people start using these notebooks more. Which brings me to my main question:

**How do we plan to keep this Colab and others up-to-date when we want to change some of the "skeleton" setup (e.g. the Google Drive mounting, autoreload configuration, code cloning and copying)**

A first possible part of the answer is: "can we extract part of this skeleton code into a mini-library within Flax? like flax_notebooks.py"? (This file would be somehow loaded first, imported, and *not* impacted by the auto-reload/Google Drive logic)

If we're not able to extract all of the skeleton code into a library, then we may want to consider this through the prism of @jheek's formal `examples/skeleton` proposal (that sets up best practices/patterns and nothing else).
Then we could consider telling people: "if you change anything in these files or in these parts of the files, then you must also get approval for change in the skeleton and have maintainers update their examples accordingly"

This response became long, and we probably need a separate GitHub issue to discuss the `examples/skeleton` proposal (does that exist?)

I think there's a bug though, (sorry GitHub didn't let me add inline comments on the ipynb file because it's too big) -- you have code that loops over 3 different optimizer momentum values but it doesn't look like you updated the config so I think you're running the same training code for all 3?

To answer your concrete questions,

> a. Do you agree mount_gdrive should default to yes? The current setup has the advantage to run end-to-end without any user interaction (authenticating for Drive requires manually copying a auth token).

It's good to not require Google Drive token for initial exploratory runs. (One small downside is that you can still edit `mnist_lib.py` and you can still press "save" in Colab which would clone the /Colab/ (but not your mnist_lib) into Drive, so users may be surprised. Maybe there's a way to detect that? Like could we `chmod` the file to be read-only when you're not using Drive? Would Colab then indeed make the files read-only in the editor?)

> b. Similar question and reasoning for upload_data dropdown.

Yes, same -- don't have people upload their results unless they know they're doing that.

> c. Do you agree we should commit the notebook with the outputs? It makes it rather large (~10M) mainly because of TensorBoard - that won't display when viewed on Github (link), but it shows fine with on Colab (link). Also, including the outputs might discourage users from actually running it (?).

I don't think it's a problem if people don't run the Colab -- if it's good enough to just browse that's a feature not a bug. It's nice to see the TB plots inline in Colab.

The problem is simply size -- 10M per file is indeed quite large and gets close to the limit of what's not good to store directly in Git (as any time anyone clones the repo they'll clone the entire history of edits which means approximately 10M * number_of_edits_per_colab_evet * number_of_colabs. See [this](https://stackoverflow.com/questions/3230721/max-file-size-and-max-repository-size-in-git#:~:text=If%20your%20repository%20exceeds%201GB,on%20working%20with%20large%20files.) where they suggest no more than 1GB per repo. This would put us dangerously close to that limit.

Why is it 10M? Is it because of the scalars stored for metrics or some TB overhead? I do think 10M is too much.

**Comment by andsteing** [comment]
Hi Avital

Thanks for your additional detailed comments!

As for your main question : How can we keep these Colabs in sync? A library would be one approach, but it has the disadvantage of more indirection (i.e. need to open the library files if you want to understand what is going on), and we would still need to keep the magic cells in sync that load the library.

What do you think about the following solution?

We have a `skeleton.ipynb` notebook some where (e.g. in `dev/`) and we have some utility `sync_notebooks.py` that automatically syncs any cells that are identical with a cell in `skeleton.ipynb`. This way it would suffice to update a single cell in the `skeleton.ipynb` and all notebooks when one of its cells are updated. We can use the same utility script (with a different flag) to compare cells and fail if there is a difference (for an automated check). If the `skeleton.ipynb` becomes unwieldy we can still separate logic into a library...

If this looks good to you, I will add the skeleton & utilities when I check in the second notebook as part of #480.

`chmod`ding the files kind of nullifies the purpose of the editor :) Instead, I added a `DISCLAIMER` to the files opened in the editor to make it very clear whether changes will be persisted:

```
## WARNING : Editing in VM - changes lost after reboot!!
########################################################
```

vs.

```
## Note : Editing in your Google Drive, changes will persist.
#############################################################
```

Other than fixing the `config.momentum = momentum` bug, I also separated MNIST-specific variables into their own cell so we can use the setup cell as a skeleton later on, added a standard "Open in Colab" button, and added some explanation as to the merits of running the Colab instead of simply reading it.

Removing the TensorBoard output reduced the overall notebook size from 11.3M to 111k (mostly PNGs) - it would have been nice to include it, but at these costs it's definitely not worth it...

Note: If you want you can review the notebook here : https://github.com/andsteing/flax/pull/3
(Not sure how I can install the ReviewNB app in the `google/flax` repository...)

**Review by jheek — APPROVED** [review]

**Pull request merged**
